sidechain/consensus/common/src/header_db.rs

```diff
@@ -0,0 +1,70 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+use itp_types::H256;
+use its_primitives::{traits::Header as HeaderT, types::header::SidechainHeader};
+use std::{collections::HashMap, hash::Hash as HashT};
+
+// Normally implemented on the Client in substrate
+pub trait HeaderDbTrait {
+	type Header: HeaderT;
+	/// Retrieves Header for the corresponding block hash.
+	fn header(&self, hash: &H256) -> Option<Self::Header>;
+}
+
+/// A mocked Header Database which allows you to take a Block Hash and Query a Block Header.
+pub struct HeaderDb<Hash, Header>(pub HashMap<Hash, Header>);
+impl<Hash, Header> HeaderDb<Hash, Header>
+where
+	Hash: PartialEq + Eq + HashT + Clone,
+	Header: Clone,
+{
+	pub fn new() -> Self {
+		Self(HashMap::new())
+	}
+
+	pub fn insert(&mut self, hash: Hash, header: Header) {
+		let _ = self.0.insert(hash, header);
+	}
+}
+
+impl<Hash, Header> From<&[(Hash, Header)]> for HeaderDb<Hash, Header>
+where
+	Hash: HashT + Eq + Copy + Clone,
+	Header: Copy + Clone,
+{
+	fn from(items: &[(Hash, Header)]) -> Self {
+		let mut header_db = HeaderDb::<Hash, Header>::new();
+		for item in items {
+			let (hash, header) = item;
+			header_db.insert(*hash, *header);
+		}
+		header_db
+	}
+}
+
+impl<Hash, Header> HeaderDbTrait for HeaderDb<Hash, Header>
+where
+	Hash: PartialEq + HashT + Into<H256> + From<H256> + std::cmp::Eq + Clone,
+	Header: HeaderT + Clone + Into<SidechainHeader>,
+{
+	type Header = SidechainHeader;
```

Comment:
You can remove the `Into<SidechainHeader>` and just use:

`type Header = Header`

Reply:
This is needed or else you get this
```rust
error[E0277]: the trait bound `SidechainHeader: From<Header>` is not satisfied
  --> sidechain/consensus/common/src/header_db.rs:42:8
   |
42 |         Some(header.clone().into())
   |              ^^^^^^^^^^^^^^ ---- required by a bound introduced by this call
   |              |
   |              the trait `From<Header>` is not implemented for `SidechainHeader`
   |
   = note: required for `Header` to implement `Into<SidechainHeader>`
help: consider extending the `where` clause, but there might be an alternative better way to express this requirement
   |
36 |     Header: HeaderT + Clone, SidechainHeader: From<Header>
   |                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For more information about this error, try `rustc --explain E0277`.
```

Reply:
The thing is that you don't need `.into()` when `Header = SidechainHeader`. Or a different question, why do you assume that the `Header` and `Sidechainheader` are not effectively the same type?